examples/control-vector-generator/control-vector-generator.cpp

```diff
@@ -0,0 +1,549 @@
+    float* result = new float[cb_data.n_embd * cb_data.n_embd];
+    std::memset(result, 0, cb_data.n_embd * cb_data.n_embd * sizeof(float));
+    for (size_t i = 0; i < cb_data.n_embd; i++) {
+        for (size_t j = 0; j < cb_data.n_embd; j++) {
+            float sum = 0.0f;
+            for (size_t k = 0; k < cb_data.n_tokens; k++) {
+                sum += cb_data.v_diff[idx][i * cb_data.n_tokens + k] * cb_data.v_diff[idx][j * cb_data.n_tokens + k];
+            }
+            result[i * cb_data.n_embd + j] = sum;
+        }
+    }
+    return result;
+}
+
+// TODO translate to ggml
+static void normalize_inplace(std::vector<float> & vec) {
+    // inefficient(?) norm computation
+    float norm = 0.0f;
+    for (const float& val : vec) {
+        norm += val * val;
+    }
+    norm = std::sqrt(norm);
+    for (float& val : vec) {
+        val /= norm;
+    }
+}
+
+// TODO translate to ggml
+static std::vector<float> mul_mat(const float * mat, const std::vector<float> & vec, size_t dim) {
+    std::vector<float> result(dim, 0.0f);
+    for (size_t i = 0; i < dim; ++i) {
+        for (size_t j = 0; j < dim; ++j) {
+            result[i] += mat[i * dim + j] * vec[j];
+        }
+    }
+    return result;
+}
+
+// TODO translate to ggml
+static std::vector<float> power_iteration(callback_data & cb_data, const float * matrix, int maxIterations = 1000, float tolerance = 1e-8) {
+    std::vector<float> b_tensor = std::vector<float>();
+    
+    // random vector gen/norm
+    std::default_random_engine generator(static_cast<unsigned int>(std::time(0)));
+    std::uniform_real_distribution<float> distribution(0.0, 1.0);
+    for (int i = 0; i < cb_data.n_embd; ++i) {
+        b_tensor.push_back(distribution(generator));
+    }
+    normalize_inplace(b_tensor);
+
+    for (int iter = 0; iter < maxIterations; ++iter) {
+
+        // store the previous one so we can check for convergence
+        std::vector<float> b_prev_tensor = b_tensor;
+
+        // matrix multiplication and renormalize
+        b_tensor = mul_mat(matrix, b_tensor, cb_data.n_embd);
+        normalize_inplace(b_tensor);
+
+        // convergence check
+        float diff = 0.0;
+        for (int i = 0; i < cb_data.n_embd; ++i) {
+            diff += std::pow(b_tensor[i] - b_prev_tensor[i], 2);
+        }
+        if (std::sqrt(diff) < tolerance) {
+            break;
+        }
+    }
+
+    return b_tensor;
+}
+
+// TODO translate to ggml
+static void pca(callback_data & cb_data) {
+    size_t n_threads = 8;
+    int    n_layers  = cb_data.v_diff.size();
+    std::vector<std::thread> threads;
+    cb_data.v_final.reserve(n_layers);
+    auto worker_function = [&](int worker_id) {
+        for (int il = worker_id; il < n_layers; il += n_threads) {
+            float * matrix = square_diff(cb_data, il);
+            std::vector<float> eigenvector = power_iteration(cb_data, matrix);
+            cb_data.v_final[il] = &eigenvector[0];
```

Comment:
This is storing a pointer to data allocated by a local that will be freed at the end of the scope. Is this intended?

Reply:
No it's not, I'm trying to figure why the output file has all vectors to 0; maybe due to this

Reply:
Oh, yeah, that would explain a lot. Whoops.

Reply:
Thanks for pointing that out. I replaced it with malloc & memcpy. Not the best way to do but I'll clean up once it works.